libsolidity/ast/ASTUtils.cpp

```diff
@@ -107,4 +107,24 @@ Type const* type(VariableDeclaration const& _variable)
 	return _variable.annotation().type;
 }
 
+std::optional<bigint> contractStorageSizeUpperBound(ContractDefinition const& _contract, VariableDeclaration::Location _location)
+{
+	solAssert(_location == VariableDeclaration::Location::Unspecified || _location == VariableDeclaration::Location::Transient);
+
+	bigint size = 0;
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts)
+		for (VariableDeclaration const* variable: contract->stateVariables())
+			if (
+				!(variable->isConstant() || variable->immutable()) &&
+				variable->referenceLocation() == _location
+			)
+			{
+				size += variable->annotation().type->storageSizeUpperBound();
+				if (size >= bigint(1) << 256)
+					return std::nullopt;
```

Comment:
I wanted to preserve the behavior of stopping as soon the limit is exceeded. It seems a bit weird with the optional, but I guess it's better than returning a negative value or something like that.

Reply:
I don't think this micro-optimization is all that important. It may actually be detrimental because we're saving a few iterations in a very rare corner case at the cost of extra overhead in all cases.

It probably has very little impact either way, so my suggestion would be to go for simpler code and remove it.